src/storage/createIndexedDBStorage.ts

```diff
@@ -0,0 +1,550 @@
+	objectStore: IDBObjectStore,
+	filter?: SimpleFilter<any>,
+	secondFilter?: SimpleFilter<any>,
+) {
+	let returnValue: {
+		keyRange?: IDBKeyRange,
+		index?: IDBIndex,
+		firstFilterUsed?: boolean,
+		secondFilterUsed?: boolean
+	} = {};
+	const valueOne = filter && filter.value;
+	if (filter && typeof valueOne !== 'undefined') {
+		const pathOne = getKey(filter);
+		if (pathOne) {
+			if (state.indices.has(pathOne) && objectStore.indexNames.contains(pathOne)) {
+				const index = objectStore.index(pathOne);
+				let firstFilterUsed = false;
+				let secondFilterUsed = false;
+				if (filter.filterType === FilterType.EqualTo) {
+					returnValue = { keyRange: IDBKeyRange.only(filter.value), index, firstFilterUsed: true };
+				}
+				else {
+					let lowerBound: any;
+					let upperBound: any;
+					let lowerInclusive = false;
+					let upperInclusive = false;
+					if (filter.filterType === FilterType.LessThan || filter.filterType === FilterType.LessThanOrEqualTo) {
+						upperBound = valueOne;
+						upperInclusive = filter.filterType === FilterType.LessThanOrEqualTo;
+						firstFilterUsed = true;
+					}
+					else if (filter.filterType === FilterType.GreaterThan || filter.filterType === FilterType.GreaterThanOrEqualTo) {
+						lowerBound = valueOne;
+						lowerInclusive = filter.filterType === FilterType.GreaterThanOrEqualTo;
+						firstFilterUsed = true;
+					}
+
+					const valueTwo = secondFilter && secondFilter.value;
+					if (secondFilter && getKey(secondFilter) === pathOne && typeof valueTwo !== 'undefined') {
+						if ((secondFilter.filterType === FilterType.LessThan ||
+							secondFilter.filterType === FilterType.LessThanOrEqualTo) && typeof upperBound === 'undefined') {
+							upperBound = valueTwo;
+							upperInclusive = secondFilter.filterType === FilterType.LessThanOrEqualTo;
+							secondFilterUsed = true;
+						}
+						else if ((secondFilter.filterType === FilterType.GreaterThan ||
+							secondFilter.filterType === FilterType.GreaterThanOrEqualTo) && typeof lowerBound === 'undefined') {
+							lowerBound = valueTwo;
+							lowerInclusive = secondFilter.filterType === FilterType.GreaterThanOrEqualTo;
+							secondFilterUsed = true;
+						}
+					}
+
+					returnValue = {
+						keyRange: createKeyRange(lowerBound, lowerInclusive, upperBound, upperInclusive),
+						index,
+						firstFilterUsed,
+						secondFilterUsed
+					};
+				}
+			}
+		}
+	}
+
+	return returnValue;
+}
+
+const createIndexedDBStorage: IndexedDBStorage = createInMemoryStorage.mixin({
+	mixin: {
+		fetch(this: Storage<{}, {}>, query?: Query<{}>): Promise<{}[]> {
+			const state = instanceStateMap.get(this);
+			let resolveTotalLength: (totalLength: number) => void;
+			let rejectTotalLength: (error: any) => void;
+			const totalLength = new Promise((resolve, reject) => {
+				resolveTotalLength = resolve;
+				rejectTotalLength = reject;
+			});
+			const fetchResult: FetchResult<{}> = <any> state.dbInitializationPromise.then(
```

Comment:
I've broken this up, as well as a few other sections, into more atomic functions.